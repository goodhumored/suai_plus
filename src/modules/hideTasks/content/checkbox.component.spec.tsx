import { cleanup, fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import React from "react";
import { faker } from "@faker-js/faker";
import { HiddenTasksRepositoryMock } from "../../../__test__/__mocks__/hiddenTasksRepository.mock";
jest.mock("./hiddenTasks.repository", () => ({
  HiddenTasksRepository: HiddenTasksRepositoryMock
}));

import HideTaskCheckbox from "./checkbox.component";

describe("HideTaskCheckbox test suit", () => {
  afterEach(cleanup);
  describe("Rendering", () => {
    describe("should be unchecked", () => {
      it("renders well", () => {
        jest
          .spyOn(HiddenTasksRepositoryMock, "isHiddenTask")
          .mockResolvedValue(false);
        const component = renderer.create(
          <HideTaskCheckbox
            onchange={jest.fn()}
            taskId={faker.datatype.number()}
          />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
      it("componentDidMount should be called", () => {
        jest
          .spyOn(HiddenTasksRepositoryMock, "isHiddenTask")
          .mockResolvedValue(false);
        const component = renderer.create(
          <HideTaskCheckbox
            onchange={jest.fn()}
            taskId={faker.datatype.number()}
          />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        expect(HiddenTasksRepositoryMock.isHiddenTask).toHaveBeenCalled();
      });
    });
    it("should be checked", () => {
      jest
        .spyOn(HiddenTasksRepositoryMock, "isHiddenTask")
        .mockResolvedValue(true);
      const component = renderer.create(
        <HideTaskCheckbox
          onchange={jest.fn()}
          taskId={faker.datatype.number()}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      expect(HiddenTasksRepositoryMock.isHiddenTask).toHaveBeenCalled();
    });
  });
  describe("click", () => {
    it("checkbox should work", () => {
      const { container } = render(
        <HideTaskCheckbox
          onchange={jest.fn()}
          taskId={faker.datatype.number()}
        />
      );
      const checkbox = container.querySelector("input") as HTMLInputElement;
      fireEvent.click(checkbox!);
      expect(checkbox?.checked).toBeTruthy();
    });
    it("should send new value to repository", () => {
      const taskId = faker.datatype.number();
      jest.spyOn(HiddenTasksRepositoryMock, "update");
      const { container } = render(
        <HideTaskCheckbox onchange={jest.fn()} taskId={taskId} />
      );
      const checkbox = container.querySelector("input") as HTMLInputElement;
      fireEvent.click(checkbox!);
      expect(checkbox?.checked).toBeTruthy();
      expect(HiddenTasksRepositoryMock.addHiddenTask).toHaveBeenLastCalledWith(
        taskId
      );
      fireEvent.click(checkbox!);
      expect(checkbox?.checked).toBeFalsy();
      expect(
        HiddenTasksRepositoryMock.removeHiddenTask
      ).toHaveBeenLastCalledWith(taskId);
    });
  });
});
