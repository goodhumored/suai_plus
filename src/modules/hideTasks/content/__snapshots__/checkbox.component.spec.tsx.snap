// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HideTaskCheckbox test suit Rendering should be checked 1`] = `
<input
  checked={false}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`HideTaskCheckbox test suit Rendering should be unchecked componentDidMount should be called 1`] = `
<input
  checked={false}
  onChange={[Function]}
  type="checkbox"
/>
`;

exports[`HideTaskCheckbox test suit Rendering should be unchecked renders well 1`] = `
<input
  checked={false}
  onChange={[Function]}
  type="checkbox"
/>
`;
